internal/db/bootstrap_logs_test.go

```diff
@@ -0,0 +1,161 @@
+	c.Assert(err, qt.IsNil)
+
+	c.Assert(logs2, qt.HasLen, 1)
+	c.Assert(logs2[0].LineNumber, qt.Equals, 1)
+	c.Assert(logs2[0].LogLine, qt.Equals, "Creating Juju controller \"diglett2\" on the-most-amazing-cloud")
+}
+
+func (s *dbSuite) TestBootstrapLogs_QueryBootstrapLogs(c *qt.C) {
+	ctx := c.Context()
+
+	err := s.Database.Migrate(ctx)
+	c.Assert(err, qt.IsNil)
+
+	// Query where the job doesn't exist
+	jobIdThatDoesntExist := uuid.New()
+	_, err = s.Database.QueryBootstrapLog(ctx, jobIdThatDoesntExist, 0)
+	c.Assert(err, qt.ErrorMatches, "job not found")
+
+	// Add job to reference
+	jobId, err := s.Database.AddJob(ctx, "test-job")
+	c.Assert(err, qt.IsNil)
+
+	// Query with no logs
+	_, err = s.Database.QueryBootstrapLog(ctx, jobId, 0)
+	c.Assert(err, qt.ErrorMatches, "not found")
+	// Query with one log
+	err = s.Database.AddBootstrapLog(ctx, jobId, "Creating Juju controller \"diglett\" on the-most-amazing-cloud")
+	c.Assert(err, qt.IsNil)
+
+	loggies, err := s.Database.QueryBootstrapLog(ctx, jobId, 0)
+	c.Assert(err, qt.IsNil)
+	c.Assert(loggies[0], qt.Equals, "Creating Juju controller \"diglett\" on the-most-amazing-cloud")
+
+	// Query with two logs, offset 0
+	err = s.Database.AddBootstrapLog(ctx, jobId, "Fetching Juju agent binaries")
+	c.Assert(err, qt.IsNil)
+
+	loggies, err = s.Database.QueryBootstrapLog(ctx, jobId, 0)
+	c.Assert(err, qt.IsNil)
+	c.Assert(loggies[0], qt.Equals, "Creating Juju controller \"diglett\" on the-most-amazing-cloud")
+	c.Assert(loggies[1], qt.Equals, "Fetching Juju agent binaries")
+
+	// Query with two logs, offset 1
+	loggies, err = s.Database.QueryBootstrapLog(ctx, jobId, 1)
+	c.Assert(err, qt.IsNil)
+	c.Assert(loggies[0], qt.Equals, "Fetching Juju agent binaries")
+
+	// Query with two logs, offset 2 (equal to the amount of logs)
+	_, err = s.Database.QueryBootstrapLog(ctx, jobId, 2)
+	c.Assert(err, qt.ErrorMatches, ".*offset cannot be greater than or equal to the amount of logs.*")
+}
+
+// This test is a behaviour check, that is, we want to see that our queued
+// locking for the bootstrap_logs table does indeed wait and prevent writes whilst
+// it is locked.
+func (s *dbSuite) TestBootstrapLogs_lockBootstrapLogs(c *qt.C) {
+	ctx := c.Context()
+
+	err := s.Database.Migrate(ctx)
+	c.Assert(err, qt.IsNil)
+
+	jobId, err := s.Database.AddJob(ctx, "test-job")
+	c.Assert(err, qt.IsNil)
+
+	finishTransaction := make(chan bool)
+	lockAcquired := make(chan bool)
+
+	go func() {
+		// Simulate a "AddBootstrapLog" call utilising the lockBootstrapLogs func.
+		// This enables us to see that locking the table does indeed prevent
+		// other AddBootstrapLog calls.
+		err := s.Database.Transaction(func(d *db.Database) error {
+			err := db.LockBootstrapLogs(d)
+			if err != nil {
+				return err
+			}
+
+			close(lockAcquired)
+
+			<-finishTransaction
+			return nil
+		})
+
+		c.Assert(err, qt.IsNil)
+	}()
+
+	<-lockAcquired
+
+	// Normal table locks do not support NOWAIT, so this is queue of INSERTS.
```

Comment:
I see on [this page](https://www.postgresql.org/docs/current/sql-lock.html) that it seems NOWAIT is supported for table locks, 
>LOCK TABLE obtains a table-level lock, waiting if necessary for any conflicting locks to be released. If NOWAIT is specified, LOCK TABLE does not wait to acquire the desired lock: if it cannot be acquired immediately, the command is aborted and an error is emitted

You could simplify your test by starting with a raw SQL query that locks the table with nowait, then call `db.AddBootstrapLog()` which should fail. As discussed this is just proving to ourselves that these locks work the way we expect.

Reply:
Ooh. Yeah I'll add that soon, would prefer to see an error. Perhaps I added it incorrectly earlier... and instead of checking further, reasoned that we would likely want to queue it. I'll add NOWAIT for the test but leave our implementation without NOWAIT.

Reply:
>I'll add NOWAIT for the test but leave our implementation without NOWAIT.

I edited my comment above, but if you can do that and it works then awesome. If not then no biggie.

Reply:
Yeah I can add it to just our test, not a problem. I think I'm convinced it's worth keeping just to demonstrate to ourselves that it works as intended.